game.js

```diff
@@ -63,6 +74,93 @@ document.querySelector('.restart').addEventListener('click', () => {
         cell.classList.remove('disabled', 'x', 'o')
     })
 
+    game.xTurn = true
+    game.xState = []
+    game.oState = []
+})
+
+document.querySelector('.return-one-step').addEventListener('click', () => {  
+    const xLength = game.xState.length;
+    const oLength = game.oState.length;
+    const lastXCell = document.querySelector(`[data-value = '${game.xState[xLength -1]}']`)
+    const lastOCell = document.querySelector(`[data-value = '${game.oState[oLength -1]}']`)
+ if (xLength > 0){
+    if (game.xTurn) { 
+     lastOCell.classList.remove('o', 'disabled');
+     game.oState.pop();
+    game.xTurn = false;
+ }
+ else{
+     lastXCell.classList.remove('x', 'disabled');
+     game.xState.pop();
+     game.xTurn = true;
+ }}
+ })
+
+ document.querySelector('.show-record').addEventListener('click', () => {
+ const scoresArr = [];
+
+ for (const score in game.winnings) {
+  scoresArr.push([score, game.winnings[score]]);
+ }
+ if (scoresArr.length > 0){
+scoresArr.sort((a , b) =>  a[1] - b[1])
+alert(`The fastest game was game ${scoresArr[0][0]} with ${scoresArr[0][1]} moves`);
+}
+else {
+    alert("No winnings yet")
+}
+})
+
+document.querySelector('.save-game').addEventListener('click', function(){
+   if(game.xState.length){
+        game.savedGame.states.push(game.xState, game.oState);
+        game.savedGame.xTurn = game.xTurn;
+       this.disabled = true;
+    }
+    else{
+        alert("The board still empty")
+    }
+})
+
+document.querySelector('.load-game').addEventListener('click', () =>{
+if(game.savedGame.states.length > 0){
+        game.xState = game.savedGame.states[0];
+        game.oState = game.savedGame.states[1];
+
+ document.querySelectorAll('.grid-cell').forEach(function(cell){
+
+return game.xState.map((state) =>{
+    if(cell.dataset.value === state){
+cell.classList.add('x');
+    }
+else{
+    cell.classList.remove('x');
+}
+});
+});
+
+document.querySelectorAll('.grid-cell').forEach(function(cell){
+      return game.oState.map((state) =>{
+            if(cell.dataset.value === state){
+        cell.classList.add('o');
+            }
+            else{
+                cell.classList.remove('o');
+            }
+        })
+            })
+            game.xTurn = game.savedGame.xTurn;
+        }
+            document.querySelector('.save-game').disabled = false;   
+})
+
+document.querySelector('.restart-after-gameover').addEventListener('click', () => {
```

Comment:
why you need another restart event . you can use the same for both buttons (and give them both the same selector)

Reply:
I thought it'll make sence to seperate them because I dont need all the actions in both cases (the first one is for restarting the same game and the secont is to start a new one)

Reply:
so make if condition inside the eventListener rather then make the common logic twice

Reply:
OK! got it.